sound/soc/intel/haswell/sst-haswell-dsp.c

```diff
@@ -304,7 +312,7 @@ static int hsw_set_dsp_D0(struct sst_dsp *sst)
 	int tries = 10;
 	u32 reg, fw_dump_bit;
 
-	/* Disable core clock gating (VDRTCTL2.DCLCGE = 0) */
+	/* Disable core clock gating and trunk clock gating */
```

Comment:
for the D3->D0 transition, you are setting the D0 state before powering up the memories

````
4. Power up necessary SRAM and wait at least for 18 clock cycles for every
    bank you have powered up
````

You are also combining steps 6 and 7

````
 6. Restore MCLK (clkctl.smos, disabled in D3 entry point 4)
    7. Stall and reset core, set CSR
````

You are missing the last step

````
    11.Unstall core
````

also the hw_sleep() makes no sense, you are doing this:

````c
/* put DSP into reset and stall */
	sst_dsp_shim_update_bits(sst, SST_CSR,
		SST_CSR_24MHZ_LPCS | SST_CSR_RST | SST_CSR_STALL,
		SST_CSR_RST | SST_CSR_STALL | SST_CSR_24MHZ_LPCS);

	hsw_set_dsp_D3(sst);
````

but withing hw_set_dsp_D3, you also play with the stall:

````c
static void hsw_set_dsp_D3(struct sst_dsp *sst)
{
	u32 val;
	u32 reg;

	/* Disable core clock gating (VDRTCTL2.DCLCGE = 0) */
	reg = readl(sst->addr.pci_cfg + SST_VDRTCTL2);
	reg &= ~SST_VDRTCL2_DCLCGE;
	writel(reg, sst->addr.pci_cfg + SST_VDRTCTL2);

	/* Stall and reset core */
	sst_dsp_shim_update_bits_unlocked(sst, SST_CSR,
					  SST_CSR_RST | SST_CSR_STALL,
					  SST_CSR_RST | SST_CSR_STALL);
````

Reply:
> for the D3->D0 transition, you are setting the D0 state before powering up the memories
> 
> ```
> 4. Power up necessary SRAM and wait at least for 18 clock cycles for every
>     bank you have powered up
> ```

The necessary SRAM is enabled dynamically later when the FW module is reloaded, at this point here, we don't have information about which SRAM blocks are necessary. You are right, we not are following up the sequence exactly. But if we say none block is necessary here, then it's fine?

> 
> You are also combining steps 6 and 7

They are not combined, but maybe the fisrt CSR_STALL is superfluous?
```
	/* stall DSP core, set clk to 192/96Mhz */ //Keyon: The CSR_STALL here might be superfluous? 
	sst_dsp_shim_update_bits_unlocked(sst,
		SST_CSR, SST_CSR_STALL | SST_CSR_DCS_MASK,
		SST_CSR_STALL | SST_CSR_DCS(4));

	/* Set 24MHz MCLK, prevent local clock gating, enable SSP0 clock */ //Keyon: 6.Restore MCLK (clkctl.smos)
	sst_dsp_shim_update_bits_unlocked(sst, SST_CLKCTL,
		SST_CLKCTL_MASK | SST_CLKCTL_DCPLCG | SST_CLKCTL_SCOE0,
		SST_CLKCTL_MASK | SST_CLKCTL_DCPLCG | SST_CLKCTL_SCOE0);

	/* Stall and reset core, set CSR */ //Keyon: 7. Stall and reset core, set CSR
	sst_dsp_shim_update_bits_unlocked(sst, SST_CSR,
					  SST_CSR_RST | SST_CSR_STALL,
					  SST_CSR_RST | SST_CSR_STALL);
```


> 
> ```
>  6. Restore MCLK (clkctl.smos, disabled in D3 entry point 4)
>     7. Stall and reset core, set CSR
> ```
> 
> You are missing the last step
> 
> ```
>     11.Unstall core
> ```

This is left and will do in hsw_boot(), we need unstall DSP core until FW is downloaded.

> 
> also the hw_sleep() makes no sense, you are doing this:
> 
> ```c
> /* put DSP into reset and stall */
> 	sst_dsp_shim_update_bits(sst, SST_CSR,
> 		SST_CSR_24MHZ_LPCS | SST_CSR_RST | SST_CSR_STALL,
> 		SST_CSR_RST | SST_CSR_STALL | SST_CSR_24MHZ_LPCS);
> 
> 	hsw_set_dsp_D3(sst);
> ```
> 
> but withing hw_set_dsp_D3, you also play with the stall:
> 
> ```c
> static void hsw_set_dsp_D3(struct sst_dsp *sst)
> {
> 	u32 val;
> 	u32 reg;
> 
> 	/* Disable core clock gating (VDRTCTL2.DCLCGE = 0) */
> 	reg = readl(sst->addr.pci_cfg + SST_VDRTCTL2);
> 	reg &= ~SST_VDRTCL2_DCLCGE;
> 	writel(reg, sst->addr.pci_cfg + SST_VDRTCTL2);
> 
> 	/* Stall and reset core */
> 	sst_dsp_shim_update_bits_unlocked(sst, SST_CSR,
> 					  SST_CSR_RST | SST_CSR_STALL,
> 					  SST_CSR_RST | SST_CSR_STALL);
> ```

Thanks, so adding stall and reset core here in my RFC commit looks not wise, let me keep it as it was.